README.md
# mpdev

**mpdev** is a music player daemon event watcher. It connects to the mpd socket and uses mpd's idle command to listen for player events. Whenever an event occurs, mpdev can carry out various activities using user defined hooks. The idea for doing mpdev comes from [mpdcron](https://alip.github.io/mpdcron/).


mpdev helps in bulding a database of your played tracks. Along with a script `mpdplaylist`, it can generate a playlist for mpd as per your taste and mood.

You can create scripts in $HOME/.mpdev directory. The default installation installs a script named $HOME/.mpdev/player for the uid 1000. The script does the following

1. scrobbles titles to last.fm and libre.fm. You have to create API keys by running lastfm-scrobbler and librefm-scrobbler one time
2. updates play counts in the sqlite stats.db. You can write your own script and update any external database. An example of this is in the player script which updates the Playcounttable for [RompR](https://fatg3erman.github.io/RompR/).
3. Synchronizes the ratings in the sticker (sqlite). You an write your own script and update any external database. An exampe of this is in the player script which updates the Ratingtable for [RompR](https://fatg3erman.github.io/RompR/) (MySQL). You can also do automatic rating to some default value by setting an environment variable AUTO\_RATING.

The above three are actually done by running a hook, a script named `player` in $HOME/.mpdev directory. You can put your own script named `player` in this directory. In fact mpdev can run specific hooks for specific types of mpd events. A hook can be any executable program or script. It will be passed arguments and have certain environment variables related to the song playing, available to it. Below is a list of of events and corresponding hooks that will be executed if available.

MPD EVENT|Hook script
---------|-----------
SONG_CHANGE|~/.mpdev/player
PLAY/PAUSE|~/.mpdev/playpause
STICKER_EVENT|~/.mpdev/sticker
MIXER_EVENT|~/.mpdev/mixer
OPTIONS_EVENT|~/.mpdev/options
OUTPUT_EVENT|~/.mpdev/output
UPDATE_EVENT|~/.mpdev/update
DATABASE_EVENT|~/.mpdev/database
PLAYLIST_EVENT|~/.mpdev/playlist
STORED_PLAYLIST_EVENT|~/.mpdev/stored_playlist
PARTITION_EVENT|~/.mpdev/partition
SUBSCRIPTION_EVENT|~/.mpdev/subscription
MESSAGE_EVENT|~/.mpdev/message
MOUNT_EVENT|~/.mpdev/mount
NEIGHBOUR_EVENT|~/.mpdev/neighbour
CUSTOM_EVENT|~/.mpdev/custom

The hooks are passed the following arguments

```
mpd-event      - Passed when the above events listed, apart from SONG_CHANGE happen.
player-event   - Passed when you play/pause player
playlist-event - Passed when the playlist changes
now-playing    - Passed when a song starts playing
end-song       - Passed when a song finishes playing
```

The default installation makes a copy of `/usr/libexec/mpdev/player` in `$HOME/.mpdev` for the user with uid `1000`.

The **mpdev** package also comes with `mpdev_update` and `mpdev_cleanup` programs that help in maintaining the sqlite databases `stats.db` and `sticker.db`.

## Environment Variables available to hooks

```
SONG_ID
SONG_URI
SONG_TITLE
SONG_ARTIST
SONG_ALBUM
SONG_DATE
SONG_GENRE
SONG_TRACK
SONG_DURATION
SONG_POSITION
SONG_LAST_MODIFIED
START_TIME
END_TIME
SCROBBLER_LASTFM
SCROBBLER_LIBREFM
VERBOSE
ELAPSED_TIME
PLAYER_STATE
DURATION
```

If you create the `stats` database, mpdev will update the last\_played field, play\_count fields in stats db. It will also update the song rating that you choose for the song. The ability to rate songs in mpd can be enabled by having the `sticker_file` keyword uncommented in `/etc/mpd.conf`. You will also need a mpd client that uses the mpd sticker command. One such player is `cantata`, which is available for all linux distros and Mac OSX. mpdev can also update [RompR](https://fatg3erman.github.io/RompR/) ratings and play counts and synchronize the ratings between [RompR](https://fatg3erman.github.io/RompR/) MySQL and sticker sqlite databases. Since mpdev runs in the background, it can keep updating [RompR](https://fatg3erman.github.io/RompR/), stats db play counts and history without you having to keep [RompR Web Frontend](https://fatg3erman.github.io/RompR/) running.

The sticker database can be enabled by having the followinng entry in `/etc/mpd.conf`

```
#
# The location of the sticker database.  This is a database which
# manages dynamic information attached to songs.
#
sticker_file                    "/var/lib/mpd/sticker.db"
#
```

If you want to use mpdev to update the [RompR](https://fatg3erman.github.io/RompR/) db, you need to set the following variables

```
ROMPR          - set this to any non-empty string (e.g. rompr)
MYSQL_HOST     - set this to the host on which rompr's MySQL database is running
MYSQL_PORT     - Port on which MySQL is running (usually 3306)
MYSQL_USER     - User for connecting to MySQL (e.g. rompr)
MYSQL_PASS     - Password for MYSQL_USER (e.g. romprdbpass)
MYSQL_DATABASE - Database name for rompr db (e.g. romprdb)
```

One can use [supervise](https://en.wikipedia.org/wiki/Daemontools) from the indimail-mta package to run this. The default rpm/debian installation of mpdev will that for you. But it will not enable rompr. To do that you need to run the following commands to create the above environment variables

```
$ sudo /bin/bash
# echo rompr         > /service/mpdev/variables/ROMPR
# echo 192.168.1.100 > /service/mpdev/variables/MYSQL_HOST
# echo 3306          > /service/mpdev/variables/MYSQL_PORT
# echo rompr         > /service/mpdev/variables/MYSQL_USER
# echo romprdbpas    > /service/mpdev/variables/MYSQL_PASS
# echo romprdb       > /service/mpdev/variables/MYSQL_DATABASE
# echo /home/pi      > /service/mpdev/variables/HOME
# echo /bin:/usr/bin > /service/mpdev/variables/PATH
# echo 6             > /service/mpdev/variables/AUTO_RATING
```

You need to restart mpdev for the new environment variables to be available to mpdev. To restart the mpdev daemon, you then just need to run the following command. `svc -d` stops the daemon, `svc -u` starts the daemon.

```
$ sudo svc -d /service/mpdev # this stops  the mpdev daemon
$ sudo svc -u /service/mpdev # this starts the mpdev daemon
```

If you want to do a source install and not have the supervise to run mpdev daemon, you could write a simple script and call it in a rc script during boot.

```
#!/bin/sh
while true
do
    env ROMPR=rompr \
        MYSQL_HOST=192.168.1.100
        MYSQL_PORT=3306
        MYSQL_USER=rompr
        MYSQL_PASS=romprdbpass
        MYSQL_DATABASE=romprdb \
        HOME=/home/pi \
        XDG_RUNTIME_DIR=/home/pi \
        PATH=\$PATH:/usr/bin:/bin \
        AUTO_RATING=6 \
    /usr/bin/mpdev -v
done
```

The stats database can be created running the `mpdev_update` program or by running sqlite3 program using the following SQL script.

```
CREATE TABLE IF NOT EXISTS song(
        id              INTEGER PRIMARY KEY,
        play_count      INTEGER DEFAULT 0,
        rating          INTEGER DEFAULT 0,
        uri             TEXT UNIQUE NOT NULL,
        duration        INTEGER,
        last_modified   INTEGER,
        date_added      INTEGER DEFAULT (strftime('%s','now')),
        artist          TEXT,
        album           TEXT,
        title           TEXT,
        track           TEXT,
        name            TEXT,
        genre           TEXT,
        date            TEXT,
        composer        TEXT,
        performer       TEXT,
        disc            TEXT,
        last_played     INTEGER,
        karma           INTEGER NOT NULL CONSTRAINT karma_percent CHECK (karma >= 0 AND karma <= 100) DEFAULT 50
);

CREATE INDEX rating on song(rating);
CREATE INDEX uri on song(uri);
CREATE INDEX last_played on song(last_played);
CREATE INDEX date_added on song(date_added);
CREATE INDEX last_modified on song(last_modified);
```

The scrobbler modules lastfm-scrobbler, librefm-scrobbler enables scrobbling to last.fm and libre.fm and can be enabled by running lastfm-scrobbler and librefm-scrobbler once as shown below

```
$ lastfm-scrobbler  --connect # for setting up lastfm
$ librefm-scrobbler --connect # for setting up librefm
```

After you have added the connection you should have lastfm-scrobbler.conf file in ~/.config/lastfm-scrobbler and librefm-scrobbler.conf file in ~/.config/librefm-scrobbler. These two files will have your `SESSION_KEY` and `API_KEY`. lastfm-scrobbler and librefm-scrobbler are a copy of moc-scrobbler from https://gitlab.com/pachanka/moc-scrobbler. mpdev will not do scrobbling if the `SESSION_KEY` and `API_KEY` haven't been setup.

By default mpdev will get the host name and port for mpd from MPD\_HOST and MPD\_PORT environment variables. MPD\_PASSWORD environment variable can be set to make mpdev connect to a password-protected mpd. If these environment variables aren’t set, mpdev connects to localhost on port 6600.. You can look at the logs while your songs are playing using `tail -f /var/log/svc/mpdev/current`

# Installation

## Source Installation

Take a look at BuildRequires field in mpdev.spec and Build-Depends field in debian/mpdev.dsc

```
$ ./configure --prefix=$prefix --libexecdir=$prefix/libexec/mpdev
$ make
$ sudo make install
```
## Create RPM / DEB

This is done by running the create\_rpm / create\_debian scripts. (Here `version` refers to the existing version of mpdev package)

```
$ pwd
/usr/local/src/mpdev
$ ./create_rpm
$ ls -l $HOME/rpmbuild/RPMS/x86_64/mpdev\*
-rw-rw-r--. 1 mbhangui mbhangui  290567 Feb  8 09:05 mpdev-0.1-1.1.fc31.x86_64.rpm
```

## Create debian package
```
$ pwd
/usr/local/src/mpdev
$ ./create_debian
$ ls -l $HOME/stage/mpdev*
-rw-r--r--  1 mbhangui mbhangui  558 Jul  2 19:30 mpdev.1-1.1_amd64.buildinfo
-rw-r--r--  1 mbhangui mbhangui  981 Jul  2 19:30 mpdev.1-1.1_amd64.changes
-rw-r--r--  1 mbhangui mbhangui 2916 Jul  2 19:30 mpdev.1-1.1_amd64.deb
```

## Prebuilt Binaries

Prebuilt binaries using openSUSE Build Service are available for mpdev for

* Debian (including ARM images for Debian 10 which work (and tested) for RaspberryPI (model 2,3 & 4) and Allo Sparky)
* Fedora

Use the below url for installation

https://software.opensuse.org//download.html?project=home%3Ambhangui%3Adietpi&package=mpdev


ChangeLog
*
Release 0.2 Start 02/03/2020
02/03/2020
01. Initial Release
02. added update_stats utility to maintain stats database
03. mpdev.c: set SONG_DURATION as seconds since epoch
04. player can now on its own add entries to stat database
14/07/2020
05. mpdev - handle empty playlist with no currentsong playing
06. added mpdev_update, mpdev_cleanup
15/07/2020
07. moved mpdev_update, mpdev_cleanup to libexec/mpdev
08. added mpdev_rename script
16/07/2020
09. mpdev_rename: added --query option
10. mpdev: use read instead of timeoutread
17/07/2020
11. mdpev: added START_TIME, END_TIME variables
12. player: refactored script to update ratings from rompr/sticker/stats
13. player: removed hardcored MySQL connection parameters
14. player: fixed variables getting clobbered by functions
19/07/2020
15. player: display sticker insert/updates after NewSong line
16. mpdev_update: added date_added field
17. mpdev_update: use st_mtime, instead of st_ctime for date_added field
18. mpdplaylist: use date_added field instead of last_modified
20/07/2020
19. mpdev: send status, currentsong command on player event to set
    ELAPSED_TIME
20. added rompr_hist play history utility for rompr
21. renamed rompr_hist to romprhist
21/07/2020
22. player: added get_rompr_playcount_ts() function to get playcount and
    timestamp from Playcounttable
23. player: option to run without sticker, stats enabled
24. player: create temp file for recording original start time
25. player: added error condition for getting rating from Ratingtable.
26. player: BUG. fixed rating tmp filename
27. player: display rating for RompR in logs in RompR scale
22/07/2020
28. added transfer_play
29. player: use AUTO_RATING env variable for automatic rating of unrated songs
30. create_service: fix for raspberrypi, dietpi
24/07/2020
31. mpdev_update: removed requirement of -m option for update mode
32. player: exit end-song if END_TIME is not set
33. player: fixed display playcount number in verbose output
28/07/2020
34. mpdev_cleanup.c: made -m optional
35. mpdev_update.c: added updation of Disc tag
35. player.in: added zztimestamp filed to Ratingtable
02/08/2020
36. set default values for play_count, rating, date_added fields.
37. use %mtime% tag from mpd datatabase for date_added filed instead of using stat(2) call
03/08/2020
38. mpdplaylist: added --host option to specify mpd host
39. mpdplaylist: added --shuffle option to shuffle playlist
05/08/2020
40. mpdplaylist: renamed --save option to --playlist
41. mpdev_rename: use SQL REPLACE instead of sed
06/08/2020
42. mpdplaylist: renamed --dummy option to --query
08/08/2020
42. mpdev.c, mpdev_cleanup.c: fixed usage/error strings
43. tcpopen.c: added missing return on connect() failure
09/08/2020
44. player: use unix domain socket if MYSQL_SOCKET is defined
10/08/2020
45. player: skip http urls
46. mpdev: skip submit song when starting from pause state
47. create_rpm: added comments to indicated cwd
48. player: replaced /tmp/mpdev with MPDEV_TMPDIR
49. player: exit now-playing when in pause or stop state
50. mpdev.c: set PLAYER_STATE environment variable when starting
51. mpdev.c: set ELAPSED_TIME, DURATION only when available
52. mpdev.c: fixed setting of initial_state
53. mpdev.c: fixed logic of play/pause/stop
11/08/2020
54. mpdev_update.c: adjust date_added for localtime
55. transfer_play: use mpc to play and remove playlist
14/08/2020
56. added playpause script
26/08/2020
57. player: added self-update of scripts in $HOME/.mpdev
27/08/2020
58. player: auto rate if $HOME/.mpdev/AUTO_RATING file is present
59. fixed path of stats.db
03/08/2020
60. mpdplaylist: fixed --clear option
05/08/2020
61. mpdplaylist: added --limit option
08/08/2020
62. mpdev_rename: fixed usage string
07/01/2021
03. create_services: fixed cp
04. mpdplaylist.in: added last_played IS NULL to daysnotheard query


tcpopen.c
/*
 * $Log: tcpopen.c,v $
 * Revision 1.7  2020-08-08 11:08:07+05:30  Cprogrammer
 * added missing return on connect failure
 *
 * Revision 1.6  2020-07-03 22:21:10+05:30  Cprogrammer
 * fixed SIGSEGV if a readonly string was used for host
 *
 * Revision 1.4  2018-05-29 21:47:11+05:30  Cprogrammer
 * removed call to gethostbyname() in ipv6 code
 *
 * Revision 1.3  2018-01-09 12:37:04+05:30  Cprogrammer
 * removed header hasindimail.h
 *
 * Revision 1.2  2016-06-13 14:17:37+05:30  Cprogrammer
 * close socket on connect() failure
 *
 * Revision 1.1  2015-08-20 19:03:05+05:30  Cprogrammer
 * Initial revision
 *
 *
 */
#include <netdb.h>
#include <unistd.h>
#include <netinet/in.h>
#include <arpa/inet.h>
#include <errno.h>
#include <sys/param.h>
#include <sys/types.h>
#include <sys/socket.h>
#include <sys/un.h>
#include <ctype.h>
#include "haveip6.h"
#include "fmt.h"
#include "env.h"
#include "scan.h"
#include "str.h"
#include "byte.h"
#include "subfd.h"
#include "strerr.h"
#ifdef sun
#include <sys/systeminfo.h>
#endif

#define MAXSLEEP                0
#define MAX_BUFF                300
#define SOCKBUF                 32768 /*- Buffer size used in Monkey service -*/
#define MAXNOBUFRETRY           60 /*- Defines maximum number of ENOBUF retries -*/

static unsigned sleeptime = MAXSLEEP + 1; /*- 0 for infinite connect */

static int
isnum(str)
	char           *str;
{
	char           *ptr;

	for (ptr = str; *ptr; ptr++)
		if (!isdigit((int) *ptr))
			return (0);
	return (1);
}

static int
setsockbuf(fd, option, size)
	int             fd, option, size;
{
	int             len, retrycount;

	len = size;
	for (retrycount = 0; retrycount < MAXNOBUFRETRY; retrycount++) {
		if (setsockopt(fd, SOL_SOCKET, option, (void *) &len, sizeof(int)) == -1) {
			if (errno == ENOBUFS) {
				usleep(1000);
				continue;
			}
			close(fd);
			return (-1);
		}
		break;
	}
	return (errno = 0);
}

int
tcpopen(host, service, port) /*- Thanks to Richard's Steven */
	char           *host;
	char           *service;
	int             port;
/*-
 * host    - Name or dotted decimal address of the
 *           other system / Pathname for a unix domain socket
 * service - Name of the service being requested.
 *           Can be NULL, if port > 0.
 *           it could be a service in /etc/services if port < 0
 * port    - if == 0, nothin special, use port# of the service.
 *           if < 0, use a reserved port
 *           if > 0, it is the port# of server (host-byte-order)
 */
{
	int             resvport, fd = -1, optval, retval, i;
	char           *ptr, *hostptr;
	struct servent *sp;
#if defined(LIBC_HAS_IP6) && defined(IPV6)
	struct addrinfo hints = {0}, *res = 0, *res0 = 0;
	char            serv[FMT_ULONG];
#else
	struct hostent *hp;
#ifdef HAVE_IN_ADDR_T
	in_addr_t       inaddr;
#else
	unsigned long   inaddr;
#endif
	struct sockaddr_in tcp_srv_addr;/*- server's Internet socket address */
#endif
	struct sockaddr_un unixaddr;	/*- server's local unix socket address */
	struct linger   linger;
	char            localhost[MAXHOSTNAMELEN];

	i = str_chr(host, '/');
	if (host && *host && host[i] && !access(host, F_OK)) {
		if ((fd = socket(AF_UNIX, SOCK_STREAM, 0)) == -1)
        	return -1;
    	unixaddr.sun_family = AF_UNIX;
    	str_copyb(unixaddr.sun_path, host, sizeof(unixaddr.sun_path));
    	if (connect(fd, (struct sockaddr *) &unixaddr, sizeof(struct sockaddr_un)) == -1) {
			close (fd);
        	return -1;
		}
		return(fd);
	}
#ifdef sun
	if (sysinfo(SI_HOSTNAME, localhost, MAXHOSTNAMELEN) > MAXHOSTNAMELEN)
#else
	if (gethostname(localhost, MAXHOSTNAMELEN))
#endif
		return (-1);
	if (!str_diff(host, localhost) || !str_diffn(host, "localhost", 10))
		hostptr = "localhost";
	else
		hostptr = host;
	if ((ptr = (char *) env_get("SLEEPTIME")) != (char *) 0) {
		if (isnum(ptr))
			scan_uint(ptr, &sleeptime);
		else {
			errno = EINVAL;
			return (-1);
		}
	}
#if defined(LIBC_HAS_IP6) && defined(IPV6)
	hints.ai_family = AF_UNSPEC;
	hints.ai_socktype = SOCK_STREAM;
	if (service != (char *) NULL) {
		if (port > 0)
			serv[fmt_ulong(serv, port)] = 0;
		else {
			if (isnum(service))
				str_copyb(serv, service, FMT_ULONG);
			else {
				if ((sp = getservbyname(service, "tcp")) == NULL) {
					errno = EINVAL;
					return (-1);
				}
				serv[fmt_ulong(serv, htons(sp->s_port))] = 0;
			}
		}
	} else
	if (port <= 0) {
		errno = EINVAL;
		return (-1);
	} else
		serv[fmt_ulong(serv, port)] = 0;
	if ((retval = getaddrinfo(hostptr, serv, &hints, &res0)))
		strerr_die7x(111, "tcpopen: ", "getaddrinfo: ", hostptr, ": ", serv, ":", (char *) gai_strerror(retval));
	for (fd = -1, res = res0; res && fd == -1; res = res->ai_next) {
		for (;;) {
			if (port >= 0) {
				if ((fd = socket(res->ai_family, res->ai_socktype, res->ai_protocol)) == -1)
					break; /*- Try the next address record in the list. */
			} else { /*- if (port < 0) */
				resvport = IPPORT_RESERVED - 1;
				if ((fd = rresvport_af(&resvport, res->ai_family)) < 0) /*- RFC 2292 */ {
					freeaddrinfo(res0);
					return (-1);
				}
			}
			for (errno = 0;;) {
				if (!(retval = connect(fd, res->ai_addr, res->ai_addrlen)))
					break;
				else {
#ifdef ERESTART
					if (errno == EINTR || errno == ERESTART)
#else
					if (errno == EINTR)
#endif
						continue;
					if (errno == ECONNREFUSED) {
						if (sleeptime <= MAXSLEEP) {
							(void) close(fd);
							if (sleeptime)
								(void) sleep(sleeptime);
							else
								(void) sleep(5);
							sleeptime += sleeptime;
							errno = ECONNREFUSED;
							break;
						} else {
							(void) close(fd);
							freeaddrinfo(res0);
							errno = ECONNREFUSED;
							return (-1);
						}
					}	/*- if (errno == ECONNREFUSED) */
					optval = errno;
					(void) close(fd);
					errno = optval;
					fd = -1;
					break;
				}
			} /*- for (errno = 0;;) */
	 		if (!retval || errno != ECONNREFUSED)
	 			break;
		} /*- for (;;) */
	 	if (!retval || errno != ECONNREFUSED)
	 		break;
		/*- try the next address record in list */
	} /*- for (res = res0; res && fd == -1; res = res->ai_next) */
	freeaddrinfo(res0);
#else /*- #if defined(LIBC_HAS_IP6) && defined(IPV6) */
	/*
	 * Initialize the server's Internet address structure. We'll store
	 * the actual 4-byte Internet address and the 2-byte port # below.
	 */
	byte_zero((char *) &tcp_srv_addr, sizeof(tcp_srv_addr));
	tcp_srv_addr.sin_family = AF_INET;
	if (service != (char *) NULL) {
		if (port > 0)
			tcp_srv_addr.sin_port = htons(port);	/*- caller's value */
		else {
			if (isnum(service)) {
				scan_int(service, &i);
				tcp_srv_addr.sin_port = htons(i);
			} else {
				if ((sp = getservbyname(service, "tcp")) == NULL) {
					errno = EINVAL;
					return (-1);
				} 
				tcp_srv_addr.sin_port = htons(sp->s_port);	/*- service's value */
			}
		}
	} else
	if (port <= 0) {
		errno = EINVAL;
		return (-1);
	} else
		tcp_srv_addr.sin_port = htons(port);
	/*
	 * First try to convert the hostname as the dotted decimal number.
	 * Only if that fails, call gethostbyname.
	 */
	if ((inaddr = inet_addr(hostptr)) != INADDR_NONE) /*- It's a dotted decimal */
		(void) byte_copy((char *) &tcp_srv_addr.sin_addr, sizeof(inaddr), (char *) &inaddr);
	else
	if ((hp = gethostbyname(hostptr)) == NULL)
		strerr_die5x(111, "tcpopen", "gethostbyname: ", hostptr, ": ", (char *) hstrerror(h_errno));
	else
		(void) byte_copy((char *) &tcp_srv_addr.sin_addr, hp->h_length, hp->h_addr);
	for (;;) {
		if (port >= 0) {
			if ((fd = socket(AF_INET, SOCK_STREAM, 0)) < 0)
				return (-1);
		} else { /*- if (port < 0) */
			resvport = IPPORT_RESERVED - 1;
			if ((fd = rresvport(&resvport)) < 0)
				return (-1);
		}
#if !defined(linux) && !defined(CYGWIN) && !defined(WindowsNT)
		if (!str_diffn(hostptr, "localhost", 10)) {
			optval = 1;
			if (setsockopt(fd, SOL_SOCKET, SO_USELOOPBACK, (char *) &optval, sizeof(optval)) == -1) {
				(void) close(fd);
				return (-1);
			}
		}
#endif
		/*- Connect to the server. */
		for (errno = 0;;) {
			if (!(retval = connect(fd, (struct sockaddr *) &tcp_srv_addr, sizeof(tcp_srv_addr))))
				break;
			else {
#ifdef ERESTART
				if (errno == EINTR || errno == ERESTART)
#else
				if (errno == EINTR)
#endif
					continue;
				if (errno == ECONNREFUSED) {
					if (sleeptime <= MAXSLEEP) {
						if (sleeptime)
							(void) sleep(sleeptime);
						else
							(void) sleep(5);
						sleeptime += sleeptime;
						(void) close(fd);
						errno = ECONNREFUSED;
						break;
					} else {
						(void) close(fd);
						errno = ECONNREFUSED;
						return (-1);
					}
				}	/*- if (errno == ECONNREFUSED) */
				optval = errno;
				(void) close(fd);
				errno = optval;
				return (-1);
			}
		} /*- for (errno = 0;;) */
	 	if (!retval)
	 		break;
	} /*- for (;;) */
#endif /*- #if defined(LIBC_HAS_IP6) && defined(IPV6) */
	if (fd == -1)
		return (-1);
	linger.l_onoff = 1;
	linger.l_linger = 1;
	if (setsockopt(fd, SOL_SOCKET, SO_LINGER, (char *) &linger, sizeof(linger)) == -1) {
		(void) close(fd);
		return (-1);
	} else
	if (setsockbuf(fd, SO_SNDBUF, SOCKBUF) == -1) {
		(void) close(fd);
		return (-1);
	} else
	if (setsockbuf(fd, SO_RCVBUF, SOCKBUF) == -1) {
		(void) close(fd);
		return (-1);
	}
	return (fd);
}	/*- end tcp_open */

void
getversion_tcpopen_c()
{
	static char    *x = "$Id: tcpopen.c,v 1.7 2020-08-08 11:08:07+05:30 Cprogrammer Exp mbhangui $";
	x++;
}
